### PR TITLE
Adding a check in reaction methods before calling getTextChannel()

### PR DESCRIPTION
**Pull request by DManstrator**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [x] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This PR adds a check in removing / clearing reaction methods, checking if the message was sent in a TextChannel and else tells the user that he is unable to remove reactions of others or clear them if the message is not from a TextChannel.

Before, `getTextChannel()` was called without a check throwing `This message was not sent in a text channel` if the message was from a PrivateChannel (or group) which could be misleading ([and actually was](https://discordapp.com/channels/81384788765712384/381889648827301889/684520138367827971) which is why I make that PR) in that context.

This also fixes a **bug** where if the SelfUser was passed to the function, it would still have thrown that exception even that would be allowed.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 1]
```diff
@@ -163,22 +163,23 @@ public boolean isPinned()
     @Override
     public RestAction<Void> clearReactions()
     {
-        if (!isFromType(ChannelType.TEXT))
-            throw new IllegalStateException("Cannot clear reactions from a message in a Group or PrivateChannel.");
```
Comment (gpluscb) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: This is likely intentional. Checks.check throws an IllegalArgumentException, which suggests that an illegal argument has been passed to the method. That is not the case here, so throwing an IllegalStateException is more appropriate. The same applies to the methods below.

Comment (DManstrator) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: Yeah, I just recognized this 10min ago and I already started changing that back to throw an IllegalStateException. Before I didn't noticed that I was throwing the wrong exception now.

I will also add a note to the Docs of Message about that case.

Thanks for telling me though!

**Comment by DManstrator** [comment]
Docs and code should be correct now, waiting for requested changes.

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 2]
```diff
@@ -172,13 +172,17 @@ public boolean isPinned()
...
+            throw new IllegalStateException("Cannot clear reactions from a message in a Group or PrivateChannel.");
         return getTextChannel().clearReactionsById(getId(), unicode);
     }
 
     @Nonnull
     @Override
     public RestAction<Void> clearReactions(@Nonnull Emote emote)
     {
+        if (!isFromType(ChannelType.TEXT))
+            throw new IllegalStateException("Cannot clear reactions from a message in a Group or PrivateChannel.");
```
Comment (duncte123) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: You should probably remove the group from here since bots can't be in groups

Comment (DManstrator) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: I am fully aware of this but since this was [the original output](https://github.com/DV8FromTheWorld/JDA/blob/fb0c6281a3eec4c07fc33ca1699d417077ada61b/src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java#L166-L167) and the throws section saying ["if was not sent in a TextChannel"](https://github.com/DV8FromTheWorld/JDA/blob/fb0c6281a3eec4c07fc33ca1699d417077ada61b/src/main/java/net/dv8tion/jda/api/entities/Message.java#L1190-L1192), I decided to stay with this.

However in my JavaDoc additions, I explicitly talk about private messages since it's easier to understand that than telling it has to come from a TextChannel imo.

TL;DR: I would wait for Minn's response since this is likely to be unwanted.

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 3]
```diff
@@ -1468,7 +1489,11 @@ default boolean isFromGuild()
      *             <li>{@link net.dv8tion.jda.api.Permission#MESSAGE_MANAGE Permission.MESSAGE_MANAGE}</li>
      *         </ul>
      * @throws java.lang.IllegalArgumentException
-     *         If the provided unicode emoji is null or empty.
+     *         If the provided unicode emoji is null or empty or if the provided user is null.
+     * @throws java.lang.IllegalStateException
+     *         If this message was <b>not</b> sent in a
+     *         {@link net.dv8tion.jda.api.entities.TextChannel TextChannel}
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java]: I'd say not sent in a guild to also account for other channel types such as news channels

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 4]
```diff
@@ -193,6 +197,14 @@ public boolean isPinned()
     @Override
     public RestAction<Void> removeReaction(@Nonnull Emote emote, @Nonnull User user)
     {
+        Checks.notNull(user, "User");  // to prevent NPEs
+        // check if the passed user is the SelfUser, then the ChannelType doesn't matter
+        if (user.equals(getJDA().getSelfUser()))
+            // we can safely remove that
+            return channel.removeReactionById(getIdLong(), emote);
+
+        if (!isFromType(ChannelType.TEXT))
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: These checks should be changed to isFromGuild checks

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 5]
```diff
@@ -1324,6 +1336,8 @@ default boolean isFromGuild()
     /**
      * Removes a {@link net.dv8tion.jda.api.entities.User User's} reaction from this Message using an {@link net.dv8tion.jda.api.entities.Emote Emote}.
      *
+     * <p>Please note that you <b>can't</b> remove reactions of other users if this is a private message / this message was sent in a {@link net.dv8tion.jda.api.entities.PrivateChannel PrivateChannel}!
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java]:
```suggestion
     * <p>Please note that you <b>can't</b> remove reactions of other users if this message was sent in a {@link net.dv8tion.jda.api.entities.PrivateChannel PrivateChannel}!
```

File: src/main/java/net/dv8tion/jda/api/entities/Message.java (hunk 3 above)
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java]: Current behavior doesn't mean its permanent, being vague allows us to make changes more easily.

Comment (DManstrator) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java]: How should I implement it then? As it is now or should I revert that back? Or do I get your comment wrong?

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 6]
```diff
@@ -682,15 +700,17 @@ public PrivateChannel getPrivateChannel()
     @Override
     public TextChannel getTextChannel()
     {
-        if (!isFromType(ChannelType.TEXT))
+        // this is not really safe but at the moment this is the only GuildChannel with messages
+        // and a MessageChannel can just be Text or Private so this is assumable as of now
+        if (!isFromGuild())
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: This check should be reverted.

**Comment by DManstrator** [comment]
In case it went under the radar, I re-requested a review.

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 7]
```diff
@@ -193,6 +197,14 @@ public boolean isPinned()
     @Override
     public RestAction<Void> removeReaction(@Nonnull Emote emote, @Nonnull User user)
     {
+        Checks.notNull(user, "User");  // to prevent NPEs
+        // check if the passed user is the SelfUser, then the ChannelType doesn't matter
+        if (user.equals(getJDA().getSelfUser()))
+            // we can safely remove that
+            return channel.removeReactionById(getIdLong(), emote);
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: nit: When the if body is more than one line it should use braces, the comment could be moved behind the return or use braces.

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:204]: I'd like to see this change made too.

**Comment by DV8FromTheWorld** [comment]
What is the progress on this PR? I see @MinnDevelopment has approved already.

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java (hunk 7 above)
Comment (DManstrator) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:204]: Let me change that really quick. :)
